### PR TITLE
feat(streaming): support union all for streaming query

### PR DESCRIPTION
**Pull request by chenzl25**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

Please explain **IN DETAIL** what the changes are in this PR and why they are needed:

- Support `UNION ALL` for streaming queries.
- The most critical part of streaming union is that we need to rewrite all the inputs of the union operator to contain pk while keeping all the input schema being the same.
- We add a new column `source_col` for rewriting the union operator to identify the record that came from the input source and make it a part of the union pks.
- This PR only supports `UNION ALL`, but not `UNION`. I think all the `UNION` should be transformed into `UNION ALL` by optimizer in the later PR to avoid the handcrafted streaming rewriting.
- This PR we enforce all the inputs of union with single distribution. I think we can optimize it with other distributions (like round robin) in the future.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Documentation

If your pull request contains user-facing changes, please specify the types of the changes, and create a release note. Otherwise, please feel free to remove this section.

### Types of user-facing changes

Please keep the types that apply to your changes, and remove those that do not apply.

* SQL operators

### Release note

Please create a release note for your changes. In the release note, focus on the impact on users, and mention the environment or conditions where the impact may occur.

## Refer to a related PR or issue link (optional)
https://github.com/risingwavelabs/risingwave/issues/6392

### CODE REVIEW

File: e2e_test/streaming/union.slt [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+----
+
+statement ok
+drop materialized view v;
+
+statement ok
+drop table t1;
+
+statement ok
+drop table t2;
```
Comment (TennyZhuang) [inline e2e_test/streaming/union.slt:57]:
```suggestion
drop table t2;

```

File: src/frontend/src/optimizer/plan_node/generic_derive.rs [hunk 2]
```diff
@@ -385,3 +385,29 @@ impl<PlanRef: GenericPlanRef> GenericPlanNode for Expand<PlanRef> {
         self.input.ctx()
     }
 }
+
+impl<PlanRef: GenericPlanRef> GenericPlanNode for Union<PlanRef> {
+    fn schema(&self) -> Schema {
+        self.inputs[0].schema().clone()
```
Comment (TennyZhuang) [inline src/frontend/src/optimizer/plan_node/generic_derive.rs:391]: Should we add the hidden column here?

File: src/frontend/src/optimizer/plan_node/generic_derive.rs [hunk 3]
```diff
@@ -385,3 +385,29 @@ impl<PlanRef: GenericPlanRef> GenericPlanNode for Expand<PlanRef> {
...
+        // Union all its inputs pks + source_col if exists
+        let mut pk_indices = vec![];
+        for input in &self.inputs {
+            for pk in input.logical_pk() {
+                if !pk_indices.contains(pk) {
+                    pk_indices.push(*pk);
+                }
+            }
+        }
+        if self.source_col.is_some() {
```
Comment (TennyZhuang) [inline src/frontend/src/optimizer/plan_node/generic_derive.rs:404]: Use `if let` here

File: src/frontend/src/optimizer/plan_node/mod.rs [hunk 4]
```diff
@@ -473,6 +475,7 @@ macro_rules! for_all_plan_nodes {
             , { Stream, DynamicFilter }
             , { Stream, ProjectSet }
             , { Stream, GroupTopN }
+            , { Stream, Union}
```
Comment (TennyZhuang) [inline src/frontend/src/optimizer/plan_node/mod.rs:478]:
```suggestion
            , { Stream, Union }
```

File: src/frontend/src/optimizer/plan_node/generic_derive.rs (hunk 2 above)
Comment (chenzl25) [reply src/frontend/src/optimizer/plan_node/generic_derive.rs:391]: Probably not. The input schema has already contains the hidden column and we can find them by `source_col`.

File: src/frontend/planner_test/tests/testdata/union.yaml [hunk 5]
```diff
@@ -9,6 +9,15 @@
...
       └─BatchScan { table: t2, columns: [t2.a, t2.b, t2.c], distribution: SomeShard }
+  stream_plan: |
+    StreamMaterialize { columns: [a, b, c, t1._row_id(hidden), null:Int64(hidden), 0:Int32(hidden)], pk_columns: [t1._row_id, null:Int64, 0:Int32] }
+    └─StreamUnion { all: true }
+      ├─StreamExchange { dist: Single }
+      | └─StreamProject { exprs: [t1.a, t1.b, t1.c, t1._row_id, null:Int64, 0:Int32] }
+      |   └─StreamTableScan { table: t1, columns: [t1.a, t1.b, t1.c, t1._row_id], pk: [t1._row_id], dist: UpstreamHashShard(t1._row_id) }
+      └─StreamExchange { dist: Single }
+        └─StreamProject { exprs: [t2.a, t2.b, t2.c, null:Int64, t2._row_id, 1:Int32] }
+          └─StreamTableScan { table: t2, columns: [t2.a, t2.b, t2.c, t2._row_id], pk: [t2._row_id], dist: UpstreamHashShard(t2._row_id) }
```
Comment (st1page) [inline src/frontend/planner_test/tests/testdata/union.yaml:20]: please add a `stream_dist_plan` test.

**Review by st1page — COMMENTED** [review]
Rest LGTM. We can find a better way when we move the pk rewriting only on stream plan :hot_face:

**Review by fuyufjh — APPROVED** [review]

**Pull request merged**